File: BasicMTGScraper/src/main/java/com/wimpy/core/MtgGoldfishScraper.java
package com.wimpy.core;

import com.google.gson.Gson;
import com.wimpy.dao.CardDatabaseCrudDao;
import com.wimpy.dao.entity.MtgCard;
import com.wimpy.dao.entity.MtgHistory;
import com.wimpy.model.MtgQuery;
import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

/**
 * This class contains the core logic to retrieve the card price
 */
@Component
public class MtgGoldfishScraper {


    private static final Logger log = LoggerFactory.getLogger(MtgGoldfishScraper.class);
    private final Map<String, String> cookies = new HashMap<>();

    @Value("${mtg.gold.fish.extract.price}")
    private String priceSelector;

    @Value("${mtg.gold.fish.extract.name}")
    private String nameSelector;

    @Value("${mtg.gold.fish.extract.edition}")
    private String editionSelector;

    private final CardDatabaseCrudDao cardDatabaseCrudDao;

    @Autowired
    public MtgGoldfishScraper(CardDatabaseCrudDao cardDatabaseCrudDao) {
        this.cardDatabaseCrudDao = cardDatabaseCrudDao;
        setupDefaultCookies();
    }

    /**
     * Setup the default cookies to scrap future request a lot faster
     */
    private void setupDefaultCookies() {
        try {
            cookies.putAll(
                    Jsoup.connect(
                                    "https://www.mtggoldfish.com/price/Throne+of+Eldraine/Edgewall+Innkeeper#paper")
                            .cookies(cookies)
                            .execute()
                            .cookies());
        } catch (IOException e) {
            log.error("Failed to update default cookies [errorMessage={}]", e.getMessage(), e);
        }
    }


    @Transactional
    public MtgHistory retrieveCardPrice(MtgQuery mtgGoldFishQuery) {

        String link = mtgGoldFishQuery.link();

        // If link is empty construct it yourself
        if (mtgGoldFishQuery.link().length() == 0) {
            link = constructLinkOfCard(mtgGoldFishQuery.cardName(), mtgGoldFishQuery.edition());
        }

        try {
            Connection.Response execute = Jsoup.connect(link).cookies(cookies).execute();
            // Updates cookies after every request
            cookies.putAll(execute.cookies());
            Document document = Jsoup.parse(execute.body());


            MtgHistory mtgHistory = new MtgHistory();

            String name = document
                    .select(nameSelector)
                    .text()
                    .replaceAll("&nbsp;", "");
            mtgHistory.setName(name);

            BigDecimal price = BigDecimal.valueOf(Double.parseDouble(document
                    .select(priceSelector)
                    .html()
                    .replaceAll("&nbsp;", "").substring(1)));
            mtgHistory.setPrice(price);

            mtgHistory.setLink(link);

            return mtgHistory;

        } catch (Exception e) {
            log.error("Failed to retrieve card details from website [link={}]", link, e);
            throw new RuntimeException("Could not retrieve card details");
        }

    }

    @Transactional
    public MtgHistory saveCardPrice(MtgQuery query) {
        MtgHistory mtgHistory = retrieveCardPrice(query);


        MtgHistory card = cardDatabaseCrudDao.save(mtgHistory);
        log.info("saved [entityJson={}]", new Gson().toJson(card));

        return card;
    }


    /**
     * Takes the cardName and edition and construct the link find the card
     *
     * @param cardName
     * @param edition
     * @return
     */
    private String constructLinkOfCard(String cardName, String edition) {

        String formattedEdition = specialFormatting(edition);
        String cardNameEdition = specialFormatting(cardName);

        return String.format("https://www.mtggoldfish.com/price/%s/%s#paper", formattedEdition, cardNameEdition);
    }


    private String specialFormatting(String text) {
        return text.replaceAll("[\\s]+", "+");
    }
}

File: BasicMTGScraper/src/main/java/com/wimpy/core/MtgGoldfishScraper.kt
package com.wimpy.core

import com.google.gson.Gson
import com.wimpy.core.util.MtgGoldfishExtractor
import com.wimpy.db.dao.MtgCardCrudDao
import com.wimpy.db.dao.MtgHistoryCrudDao
import com.wimpy.db.entity.MtgCard
import com.wimpy.db.entity.MtgHistory
import com.wimpy.rest.v1.model.MtgQuery
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

/**
 * This class contains the core logic to retrieve the card price
 */
@Component
open class MtgGoldfishScraper @Autowired constructor(
    private val mtgGoldfishExtractor: MtgGoldfishExtractor,
    private val mtgHistoryCrudDao: MtgHistoryCrudDao,
    private val mtgCardCrudDao: MtgCardCrudDao,
    private val scraperUtil: ScraperUtil
) {

    private val log = LoggerFactory.getLogger(MtgGoldfishScraper::class.java)

    @Transactional
    open fun retrieveCardPrice(mtgGoldFishQuery: MtgQuery): MtgHistory {
        var link = mtgGoldFishQuery.link

        //https://www.mtggoldfish.com/price/Commander 2017/Alms Collector#paper
        if (mtgGoldFishQuery.cardName.isBlank() && link.endsWith("#paper"))
            mtgGoldFishQuery.cardName = link.substringAfterLast("/").substringBefore("#paper")

        // If link is empty construct it yourself
        if (mtgGoldFishQuery.link.length == 0) {
            link = constructLinkOfCard(mtgGoldFishQuery.cardName, mtgGoldFishQuery.edition)
        }

        try {
            val document = scraperUtil.retrieveOnline(link)
            return mtgGoldfishExtractor.extractFields(document, link)
        } catch (e: Exception) {
            log.error("Failed to retrieve card details from website [link={}]", link, e)
            throw RuntimeException("Could not retrieve card details", e)
        }
    }

    @Transactional
    open fun saveCardPrice(mtgHistory: MtgHistory): MtgHistory {
        val optionalMtgCard = mtgCardCrudDao.findByName(mtgHistory.name)


        //if empty creates the card and links it up to the history
        optionalMtgCard.ifPresentOrElse({ mtgCard: MtgCard? -> mtgHistory.mtgCard = mtgCard }) {
            val entity = MtgCard()
                .setName(mtgHistory.name)
            mtgHistory.mtgCard = mtgCardCrudDao.save(entity)
        }
        val card = mtgHistoryCrudDao.save(mtgHistory)
        log.info("saved [entityJson={}]", Gson().toJson(card))
        return card
    }

    /**
     * Takes the cardName and edition and construct the link find the card
     *
     * @param cardName
     * @param edition
     * @return
     */
    private fun constructLinkOfCard(cardName: String, edition: String): String {
        val formattedEdition = specialFormatting(edition)
        val cardNameEdition = specialFormatting(cardName)
        return "https://www.mtggoldfish.com/price/${formattedEdition}/${cardNameEdition}#paper";
    }

    private fun specialFormatting(text: String): String {
        return text.replace("[\\s]+".toRegex(), "+")
    }

}